File: scripts/qa/scroll_behavior.py
import asyncio
from playwright.async_api import async_playwright

BASE_URL = "http://localhost:8080"
ROUTES = ["/", "/projects", "/developers", "/faq"]


async def assert_normal_scroll(page, route: str):
    await page.goto(BASE_URL, wait_until="domcontentloaded")
    await page.evaluate("sessionStorage.setItem('jj_intro_seen', '1')")
    await page.goto(f"{BASE_URL}{route}", wait_until="domcontentloaded")
    await page.wait_for_timeout(900)
    await page.mouse.move(560, 426)
    await page.evaluate("window.scrollTo({ top: 0, behavior: 'auto' })")
    await page.wait_for_timeout(80)

    can_scroll = await page.evaluate("document.documentElement.scrollHeight > window.innerHeight + 20")
    if not can_scroll:
        return {"route": route, "moves": [], "total": 0, "skipped": True}

    moves = []
    for _ in range(6):
        before = await page.evaluate("window.scrollY")
        await page.mouse.wheel(0, 80)
        await page.wait_for_timeout(120)
        after = await page.evaluate("window.scrollY")
        moves.append(after - before)

    positive_moves = [move for move in moves if move > 0]
    biggest_move = max(moves)
    total_move = sum(moves)

    if len(positive_moves) < 5:
        raise AssertionError(f"{route}: wheel input did not consistently scroll ({moves})")
    if biggest_move > 220:
        raise AssertionError(f"{route}: small wheel input jumped too far ({moves})")
    if total_move < 300 or total_move > 950:
        raise AssertionError(f"{route}: total scroll distance is abnormal ({total_move}; {moves})")

    return {"route": route, "moves": moves, "total": total_move}


async def main():
    async with async_playwright() as playwright:
        browser = await playwright.chromium.launch(headless=True)
        context = await browser.new_context(viewport={"width": 1121, "height": 853}, device_scale_factor=2)
        page = await context.new_page()
        results = []
        for route in ROUTES:
            results.append(await assert_normal_scroll(page, route))
        await browser.close()
        print({"ok": True, "results": results})


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/qa/scroll-behavior.mjs
import { chromium } from "playwright";

const BASE_URL = process.env.QA_BASE_URL || "http://localhost:8080";
const ROUTES = ["/", "/projects", "/developers", "/faq"];

const assertNormalScroll = async (page, route) => {
  await page.goto(`${BASE_URL}${route}`, { waitUntil: "domcontentloaded" });
  await page.waitForTimeout(900);
  await page.evaluate(() => window.scrollTo({ top: 0, behavior: "auto" }));
  await page.waitForTimeout(80);

  const scrollHeight = await page.evaluate(() => document.documentElement.scrollHeight);
  if (scrollHeight <= window.innerHeight) return;

  const moves = [];
  for (let i = 0; i < 6; i += 1) {
    const before = await page.evaluate(() => window.scrollY);
    await page.mouse.wheel(0, 80);
    await page.waitForTimeout(120);
    const after = await page.evaluate(() => window.scrollY);
    moves.push(after - before);
  }

  const positiveMoves = moves.filter((move) => move > 0);
  const biggestMove = Math.max(...moves);
  const totalMove = moves.reduce((sum, move) => sum + move, 0);

  if (positiveMoves.length < 5) {
    throw new Error(`${route}: wheel input did not consistently scroll the page (${moves.join(", ")})`);
  }

  if (biggestMove > 220) {
    throw new Error(`${route}: small wheel input jumped too far (${moves.join(", ")})`);
  }

  if (totalMove < 300 || totalMove > 950) {
    throw new Error(`${route}: total scroll distance is abnormal (${totalMove}; moves ${moves.join(", ")})`);
  }

  return { route, moves, totalMove };
};

const browser = await chromium.launch({ headless: true });
const context = await browser.newContext({ viewport: { width: 1121, height: 853 }, deviceScaleFactor: 2 });
const page = await context.newPage();

try {
  const results = [];
  for (const route of ROUTES) {
    results.push(await assertNormalScroll(page, route));
  }
  console.log(JSON.stringify({ ok: true, results }, null, 2));
} finally {
  await browser.close();
}